Alexsat/fix pix_fmt for multiple ffmpeg versions

AlexSat opened the pull request.

AlexSat committed:
Added recognizing ffmpeg versions to resolve pix_fmt argument correctly

AlexSat committed:
Remove unused go file

gabek commented on config/constants.go:
```diff
@@ -5,8 +5,8 @@ import "path/filepath"
-	// FfmpegSuggestedVersion is the version of ffmpeg we suggest.
-	FfmpegSuggestedVersion = "v4.1.5" // Requires the v
+	// FfmpegWithGeneralPixFmtVersion is the version of ffmpeg user for 'vaapi' fix_fmt value, instead of 'vaapi_vld'.
+	FfmpegWithGeneralPixFmtVersion = "v5.0" // Requires the v
```
I know we had that ffmpeg version in constants before. The thought was it would be changed over time, but that never really happened. The ffmpeg version pixel format version will never need to be changed, so maybe it should just live local to the single place it's being used.

AlexSat replied:
I check it and found usage in several places: codecs and tests, so i refactored usage

gabek commented:
Seems to have two linter errors:

```
 Error: string `vaapi` has 3 occurrences, make it a constant (goconst)
  Error: commentFormatting: put a space between `//` and comment text (gocritic)
```

gabek commented:
And a test failure:

```
--- FAIL: TestOldFFmpegVaapiCommand (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
```

AlexSat committed:
Refactor 'vaapi' to constant

AlexSat committed:
Remove disputable SetCodec method
It used in double-create-way (create codec -> get it's name -> create codec by name) in tests and replace it with code usage assigning

AlexSat committed:
Refactored usage of breaking changes ffmpeg version FfmpegWithGeneralPixFmtVersion

gabek commented:
How come you're doing version comparison instead of relying on the actual list of pixel formats? ffmpeg version testing has not been very reliable in the past. Not all builds of ffmpeg report versions the same way and it resulted in not knowing the actual version for many ffmpeg builds.

AlexSat committed:
Fix comments period

AlexSat commented:
> How come you're doing version comparison instead of relying on the actual list of pixel formats? ffmpeg version testing has not been very reliable in the past. Not all builds of ffmpeg report versions the same way and it resulted in not knowing the actual version for many ffmpeg builds.

Oh, You are right, I should use list of pixels format. I fix it by next commits (for a few days, i think)

AlexSat committed:
Get supported pixel formats from ffmpeg

AlexSat committed:
Added dot to end of comment

AlexSat commented:
@gabek fixed

AlexSat committed:
Remove panic behaviour

AlexSat committed:
Fix gosec linter stopper: fix call command for ffmpeg

AlexSat committed:
Fix imports order and grouping

gabek commented on core/transcoder/codecs.go:
```diff
@@ -23,6 +29,16 @@ type Codec interface {
 }
 
+// MockedCodec is mock for each codec to easy testing. Respect to https://github.com/anpavlov .
+type MockedCodec struct {
```
What's the mock wrapper for? Additionally it feels wrong to have test-specific code in application code.

AlexSat replied:
You are right, i should use earthfile instead host system to run tests. Fixed it

gabek commented on core/transcoder/transcoder.go:
```diff
@@ -447,8 +447,3 @@ func (t *Transcoder) SetIdentifier(output string) {
 }
-
-// SetCodec will set the codec to be used for the transocder.
-func (t *Transcoder) SetCodec(codecName string) {
```
Why was the setter removed? If anything I think we should standardize on setters instead of setting properties directly.

AlexSat replied:
Yes, I agree.

gabek commented on core/transcoder/codecs.go:
```diff
@@ -31,6 +47,14 @@ var supportedCodecs = map[string]string{
 	(&VideoToolboxCodec{}).Name(): "videotoolbox",
 }
 
+var prefferedPixelFormats = map[string][]string{
```
Think we could have each of these formats that are used for each codec as a part of the implementation of the codec itself? Much like how you can ask a codec for its name or flags.

AlexSat replied:
But I have not idea how to choose pixel format... Ffmpeg returns so many pixel formats and I dont know how to make best choice. Could you help me with it, please? Could you explain choose-pixel-fmt-way to me, please?

gabek reviewed (commented):
Awesome, thank you! I left a couple questions and thoughts.

AlexSat committed:
Remove useless CodecMock. Introduce back SetCodec method

AlexSat committed:
Added ffmpeg to unit test container

AlexSat committed:
Added comment

gabek commented:
Looks like tests are failing.
